### ISSUE TITLE
Strange behaviour in sysrepo when mandatory=true leafs of different models

### ISSUE THREAD

**Issue by vhuertas** [sysrepo/sysrepo#1466]
### Things you may try first
(put "x" in "[]" if you already tried following)
* [] Did you check if this is a duplicate issue?
* [] Did you test it on the latest sysrepo devel branch?

### Description
Two YANG models are in use: ietf-interfaces (plus all augments of ietf-ip and others) plus another YANG model defined by me (routing-indra.yang) that has augments for several ietf-models.
Using sysrepocfg I try to merge an XML file where I configure a new IPv6 address in the ietf-interfaces model and the sysrepo complains that the merge cannot be done because there is a missing field. This field has the mandatory as TRUE but it is located in the model defined by me that has nothing to do with the ietf-models.
Moreover, when I perform a merge with anothef XML file that only affects the routing-indra model, then the sysrepod report successful merging (as expected)
I don't know if it is a bug of the sysrepo that I am using a this moment or not.

### Steps to Reproduce
1. [First Step]
Install the ietf-interfaces, ietf-ip and routing-indra yang models (see the output of the sysrepoctl --list)
cuda1@cuda1:~> sysrepoctl --list
Sysrepo schema directory: /etc/sysrepo/yang/
Sysrepo data directory:   /etc/sysrepo/data/
(Do not alter contents of these directories manually)

Module Name                | Revision   | Conformance | Data Owner          | Permissions | Submodules                    | Enabled Features
-----------------------------------------------------------------------------------------------------------------------------------------------
ietf-netconf-notifications | 2012-02-06 | Installed   | root:root           | 666         |                               |
ietf-netconf               | 2011-06-01 | Installed   | root:root           | 666         |                               | writable-running candidate rollback-on-error validate startup xpath url
ietf-netconf-acm           | 2012-02-22 | Imported    |                     |             |                               |
nc-notifications           | 2008-07-14 | Installed   | root:root           | 666         |                               |
notifications              | 2008-07-14 | Installed   | root:root           | 666         |                               |
iana-if-type               | 2014-05-08 | Installed   |                     |             |                               |
ietf-interfaces            | 2014-05-08 | Installed   | root:root           | 666         |                               | if-mib
ietf-ip                    | 2014-06-16 | Installed   |                     |             |                               |
ietf-netconf-with-defaults | 2011-06-01 | Installed   |                     |             |                               |
ietf-netconf-monitoring    | 2010-10-04 | Installed   | root:root           | 666         |                               |
ietf-yang-library          | 2018-01-17 | Installed   | root:root           | 666         |                               |
ietf-yang-types            | 2013-07-15 | Installed   |                     |             |                               |
turing-machine             | 2013-12-27 | Installed   | root:root           | 666         |                               |
ietf-datastores            | 2017-08-17 | Installed   |                     |             |                               |
ietf-routing               | 2015-10-16 | Installed   | root:root           | 666         |                               |
ietf-ipv4-unicast-routing  | 2015-10-16 | Installed   |                     |             |                               |
ietf-ipv6-unicast-routing  | 2015-10-16 | Installed   |                     |             |                               |
ietf-qos-policy            | 2016-03-03 | Installed   | root:root           | 666         |                               |
ietf-qos-classifier        | 2016-03-03 | Installed   | root:root           | 666         |                               |
ietf-qos-action            | 2016-06-15 | Installed   | root:root           | 666         |                               |
ietf-access-control-list   | 2016-10-12 | Installed   | root:root           | 666         |                               |
ietf-packet-fields         | 2016-10-12 | Installed   |                     |             |                               |
ietf-inet-types            | 2013-07-15 | Installed   |                     |             |                               |
ietf-event-notifications   | 2016-10-27 | Installed   | root:root           | 666         |                               |
ydk                        | 2016-02-26 | Installed   | root:root           | 666         |                               |
routing-indra             | 2018-09-18 | Installed   | root:root           | 666         |                               | *

2. Run the command
cuda1@cuda1:~> sudo sysrepocfg -d running -l 4 --format=xml --merge=ipv6_merge.xml ietf-interfaces
[DBG] Setting log level for stderr logs to 4.
[DBG] Connecting to Sysrepo Engine.
[DBG] Connecting to socket=/var/run/sysrepod.sock
[INF] Connected to daemon Sysrepo Engine at socket=/var/run/sysrepod.sock
[DBG] Sending version-verify request.
[DBG] version-verify request sent, waiting for response.
[DBG] version-verify response received, processing.
[DBG] Sending session-start request.
[DBG] session-start request sent, waiting for response.
[DBG] session-start response received, processing.
[DBG] Sending check-enabled-running request.
[DBG] check-enabled-running request sent, waiting for response.
[DBG] check-enabled-running response received, processing.
[DBG] Loading module schema: '/etc/sysrepo/yang/ietf-interfaces@2014-05-08.yang'.
[DBG] Sending discard-changes request.
[DBG] discard-changes request sent, waiting for response.
[DBG] discard-changes response received, processing.
[DBG] Sending get-items request.
[DBG] get-items request sent, waiting for response.
[DBG] get-items response received, processing.
[DBG] No items found for xpath '/ietf-routing:*//.'
[WRN] libyang: Missing required element "max-ip-routes" in "license-params". (/routing-indra:routing-service-config/license-params)
[ERR] Input data are not valid: Missing required element "max-ip-routes" in "license-params". (/routing-indra:routing-service-config/license-params)
Errors were encountered during importing. Cancelling the operation.
[DBG] Sending session-stop request.
[DBG] session-stop request sent, waiting for response.
[DBG] session-stop response received, processing.

3.Run the command
cuda1@cuda1:~> sudo sysrepocfg -d running -l 4 --format=xml --merge=rt_srv.xml routing-indra
[sudo] password for root: 
[DBG] Setting log level for stderr logs to 4.
[DBG] Connecting to Sysrepo Engine.
[DBG] Connecting to socket=/var/run/sysrepod.sock
[INF] Connected to daemon Sysrepo Engine at socket=/var/run/sysrepod.sock
[DBG] Sending version-verify request.
[DBG] version-verify request sent, waiting for response.
[DBG] version-verify response received, processing.
[DBG] Sending session-start request.
[DBG] session-start request sent, waiting for response.
[DBG] session-start response received, processing.
[DBG] Sending check-enabled-running request.
[DBG] check-enabled-running request sent, waiting for response.
[DBG] check-enabled-running response received, processing.
[DBG] Loading module schema: '/etc/sysrepo/yang/routing-indra@2018-09-18.yang'.
[DBG] Sending discard-changes request.
[DBG] discard-changes request sent, waiting for response.
[DBG] discard-changes response received, processing.
[DBG] Sending get-items request.
[DBG] get-items request sent, waiting for response.
[DBG] get-items response received, processing.
[DBG] No items found for xpath '/ietf-routing:*//.'
[DBG] Sending get-items request.
[DBG] get-items request sent, waiting for response.
[DBG] get-items response received, processing.
[DBG] Sending get-items request.
[DBG] get-items request sent, waiting for response.
[DBG] get-items response received, processing.
[DBG] All items has been read for xpath '/ietf-interfaces:*//.'
[DBG] Sending get-items request.
[DBG] get-items request sent, waiting for response.
[DBG] get-items response received, processing.
[DBG] Sending get-items request.
[DBG] get-items request sent, waiting for response.
[DBG] get-items response received, processing.
[DBG] All items has been read for xpath '/routing-indra:*//.'
[DBG] No changes were made.
The new configuration was successfully applied.
[DBG] Sending session-stop request.
[DBG] session-stop request sent, waiting for response.
[DBG] session-stop response received, processing.

**Expected behavior:**
 when merging the ipv6_merge.xml the sysrepo debug logs should not display:
[WRN] libyang: Missing required element "max-ip-routes" in "license-params". (/routing-indra:routing-service-config/license-params)
[ERR] Input data are not valid: Missing required element "max-ip-routes" in "license-params". (/routing-indra:routing-service-config/license-params)
Errors were encountered during importing. Cancelling the operation.

**Actual behavior:**
The ipv6 address cannot be configured because of this strange problem.

### Versions
Please provide versions for software needed to reproduce the bug:
Sysrepo version: 0.7.7
libyang version: 0.16.108
netopeer2-server: 0.7.3

### Attachments

[ipv6_merge.xml.txt](https://github.com/sysrepo/sysrepo/files/3077758/ipv6_merge.xml.txt)
[logs commands.txt](https://github.com/sysrepo/sysrepo/files/3077759/logs.commands.txt)
[routing-indra@2018-09-18_yang.txt](https://github.com/sysrepo/sysrepo/files/3077760/routing-indra%402018-09-18_yang.txt)
[rt_svr.xml.txt](https://github.com/sysrepo/sysrepo/files/3077761/rt_svr.xml.txt)

**Comment by michalvasko** [comment]
Hi,
do you have *sysrepod* running when executing the *sysrepocfg* commands? If so, try it without it. Otherwise I will try to reproduce it and will see what I can do.

Regards,
Michal

**Comment by vhuertas** [comment]
Hi Michal,

I am using the version 0.7.7 from master branch.
I cannot execute the sysrepocfg without running the sysrepod because it complains that there is no subscription on the modules which I want to make the merge on.

I have also tested the same IPv6  merge using real NETCONF message with the same result and same logs with debug activated.
Do you think that this may be a bug?

Thanks for your attention

**Comment by michalvasko** [comment]
Hi,
yes, it is a bug, both in *libyang* and *sysrepocfg*. I believe I have fixed the problem in *libyang* so if you send the same XML using NETCONF, it should work as expected. However, *sysrepocfg* does not work properly with features which is causing this error. I am sorry, but you have quite a specific use-case for which you simply cannot currently use *sysrepocfg*. I may fix it later, if I have some time.

Regards,
Michal

**Comment by vhuertas** [comment]
Hi,
Thanks for your answer. Thus, Is this bug solved for libyang and therefore for sysrepo as well? Do I have to update the master source code for the last version for both libyang and sysrepo and do "make isntall"? The most urgent thing for me is to be able to use NETCONF and not sysrepocfg.
Thanks again

**Comment by michalvasko** [comment]
Hi,
using the latest *libyang* (either `devel` or `master`) should be enough. Depending on the current *libyang* version, you may need to recompile and reinstall *sysrepo* and/or *netopeer2* as well.

Regards,
Michal

**Comment by vhuertas** [comment]
After having done a pull of libyang, sysrepo and netopeer 2-server now I have problems in compilation.
Libyang is successfully compiled (make) and installed (make install). However, when doing the same with master branch of sysrepo this is the errors displayed with doing 'make':
                           ^
/home/cuda1/git/sysrepo/src/common/sr_utils.c:705:27: note: each undeclared identifier is reported only once for each function it appears in
/home/cuda1/git/sysrepo/src/common/sr_utils.c: In function ?sr_libyang_anydata_copy_value?:
/home/cuda1/git/sysrepo/src/common/sr_utils.c:1324:10: error: ?LYD_ANYDATA_LYBD? undeclared (first use in this function)
     case LYD_ANYDATA_LYBD:
          ^
/home/cuda1/git/sysrepo/src/common/sr_utils.c:1325:10: error: ?LYD_ANYDATA_LYB? undeclared (first use in this function)
     case LYD_ANYDATA_LYB:
          ^
In file included from /home/cuda1/git/sysrepo/src/common/sr_common.h:47:0,
                 from /home/cuda1/git/sysrepo/src/common/sr_utils.c:40:
/home/cuda1/git/sysrepo/src/common/sr_utils.c: In function ?sr_subtree_to_dt?:
/home/cuda1/git/sysrepo/src/common/sr_utils.c:1933:92: error: too many arguments to function ?ly_errmsg?

And now I am stuck here.... Any idea of what is happening?

Thanks again

**Comment by michalvasko** [comment]
Hi,
*sysrepo* is being compiled with some quite old `libyang.h` (specifically `tree_data.h`) without LYB format support. Best is to completely uninstall all *libyang* and *sysrepo* versions before installing new ones to different paths.

Regards,
Michal

**Comment by vhuertas** [comment]
The thing is that when you do "make install" with the last code of libyang of master branch the *.so lib files are installed in the /usr/lib64 folder instead of /usr/local/lib64.
So now I have two libyang .so versions: the old one is in /usr/local/lib64 and the new one in /usr/lib64.

When I execute "yanglint -v" it tells the old version. How do you suggest to solve this issue? May be the best should be to force the "make install" to install all the library files in the /usr/local/lib64...

Or may be to manually remove all the libyang files manually from the /usr/local/lib64 and set the LD_LIBRARY_PATH to include /usr/lib64 (now it is included but it is always chosen the old libyang version).

Regards,

**Comment by michalvasko** [comment]
Hi,
I think you should first remove **all** *libyang* versions. Then install the recent one wherever you want and then stick with it. You can change the prefix not to be `/usr/local`, naturally.

Regards,
Michal

**Comment by vhuertas** [comment]
Thanks Michal,
I managed to solve the problem and I am updated the code for all versions:

ligyang version: 1.1.5
sysrepo version: 0.7.7 (after having updated the source code the version becomes the same)
netopeer2-server: 0.7.11


However, when I try to perform the same merge as before, a new (and different) error occurs. I show you the logs that sysrepo reports (I show you only the error logs):

[ERR] libyang error: Target context does not contain schema node for the data node being duplicated (ietf-ip:ipv6).
[ERR] Failed to duplicate ietf-interfaces data tree into another context
[ERR] Failed to append data tree ietf-interfaces
[ERR] Loading dependant modules failed for routing-indra

 Do you recommend me to open a new issue of it is already open by other user?

Thanks again

**Comment by michalvasko** [comment]
Hi,
like I said previously, it did work for me when I tried with the latest versions using NETCONF. It seems there is some problem with the schemas, so you can try removing all current relevant ones (`ietf-interfaces`,  `ietf-ip`, `iana-if-type`, and your proprietary ones) and them install them again.

Regards,
Michal

**Comment by vhuertas** [comment]
OK Michal,
I will do it now and tell you how it goes.
Thanks a lot!

**Comment by vhuertas** [comment]
Dear Michal,
I have removed all the models and installed them again.
When performing the NETCONF merge message to configure the IPv6 address it appears exactly the same libyang error of the very begining. I enclose the logs:

[DBG] Loaded module ietf-interfaces: mtime sec=1555494644 nsec=692821054
[DBG] Current time: mtime sec=1555494697 nsec=494671434
[DBG] Timestamp for the model ietf-interfaces matches, ops will be skipped
[DBG] Usage count ietf-interfaces incremented (value=3)
[DBG] Loaded module ietf-interfaces: mtime sec=1555494644 nsec=692821054
[DBG] Usage count ietf-interfaces incremented (value=4)
[INF] Data file /etc/sysrepo/data/ietf-interfaces.running loaded successfully
[DBG] Switching identity to UID='0' and GID='0' (username: root).
[DBG] Commit (3/10): all modified models loaded successfully
[DBG] Skipping op for model ietf-interfaces
[DBG] Skipping op for model ietf-interfaces
[DBG] Skipping op for model ietf-interfaces
[DBG] Skipping op for model ietf-interfaces
[DBG] Skipping op for model ietf-interfaces
[DBG] Skipping op for model ietf-interfaces
[DBG] Skipping op for model ietf-interfaces
[DBG] Commit (4/10): replay of operation succeeded
[DBG] Switching identity to UID='0' and GID='0' (username: root).
[DBG] Switching identity to UID='0' and GID='0' (username: root).
[DBG] Loaded module ietf-routing: mtime sec=1555494132 nsec=706344116
[DBG] Usage count ietf-routing incremented (value=1)
[INF] Data file /etc/sysrepo/data/ietf-routing.running is empty
[DBG] Module ietf-routing has been loaded
[DBG] Dependant module ietf-routing is empty
[DBG] Usage count ietf-routing decremented (value=0)
[DBG] Data tree ietf-routing appended because of validation
[DBG] Switching identity to UID='0' and GID='0' (username: root).
[DBG] Switching identity to UID='0' and GID='0' (username: root).
[DBG] Loaded module ietf-qos-policy: mtime sec=1555494136 nsec=814299833
[DBG] Usage count ietf-qos-policy incremented (value=1)
[INF] Data file /etc/sysrepo/data/ietf-qos-policy.running loaded successfully
[DBG] Module ietf-qos-policy has been loaded
[DBG] Usage count ietf-qos-policy decremented (value=0)
[DBG] Data tree ietf-qos-policy appended because of validation
### [ERR] libyang error: Missing required element "max-ip-routes" in "license-params".
[DBG] Validation failed for ietf-interfaces module
[DBG] Switching identity to UID='0' and GID='0' (username: root).
[DBG] Switching identity to UID='0' and GID='0' (username: root).
[DBG] Loaded module tcp-acceleration-indra: mtime sec=1555494132 nsec=706344116
[DBG] Usage count tcp-acceleration-indra incremented (value=1)
[INF] Data file /etc/sysrepo/data/tcp-acceleration-indra.running is empty
[DBG] Module tcp-acceleration-indra has been loaded
[DBG] Module ietf-interfaces already loaded
[DBG] Data tree ietf-interfaces appended because of validation
[DBG] Switching identity to UID='0' and GID='0' (username: root).
[DBG] Switching identity to UID='0' and GID='0' (username: root).
[DBG] Loaded module ietf-routing: mtime sec=1555494132 nsec=706344116
[DBG] Usage count ietf-routing incremented (value=1)
[INF] Data file /etc/sysrepo/data/ietf-routing.running is empty

I don't know what I amb doing wrong to get this error whereas you don't got any error using NETCONF...
The only work around is to try to remove the mandatory=true from all the license-params leafs in routing-indra and see if it is solved. It is not the desired solution but it is the only thing I can do right now unless a solution for this bug is found.

Regards,

**Comment by vhuertas** [comment]
Michal,

Just an update: I have modified routing-indra removing the mandatory=true from all parameters and the Ipv6 address config merging NETCONF message has worked fine.

So it seems that the problem reported at the very begining is still present.

Regards,

**Comment by michalvasko** [comment]
Hi,
I have used only [these](https://github.com/sysrepo/sysrepo/files/3089177/issue.tar.gz) files and I am starting with a fresh install of *sysrepo* and *netopeer2*. I have first removed the test modules that could cause problems using
```
sysrepoctl -u -m ietf-interfaces,iana-if-type,ietf-ip
```
Then my sysrepo modules (`sysrepoctl -l`) were:
```
Module Name                | Revision   | Conformance | Data Owner          | Permissions | Submodules                    | Enabled Features
-----------------------------------------------------------------------------------------------------------------------------------------------
ietf-netconf-acm           | 2018-02-14 | Installed   | root:root           | 644         |                               |
example-module             |            | Installed   | root:root           | 666         |                               |
test-module                |            | Installed   | root:root           | 666         |                               |
referenced-data            |            | Installed   | root:root           | 666         |                               |
small-module               |            | Installed   | root:root           | 666         |                               |
info-module                |            | Installed   |                     |             |                               |
id-ref-aug                 |            | Installed   |                     |             |                               |
id-ref-base                |            | Implemented | root:root           | 666         |                               |
id-def-base                |            | Imported    |                     |             |                               |
id-ref-hello               |            | Installed   | root:root           | 666         |                               |
id-ref-hello-cont          |            | Imported    |                     |             |                               |
id-def-extended            |            | Installed   |                     |             |                               |
id-ref-main                |            | Installed   | root:root           | 666         |                               |
id-ref-imported            |            | Imported    |                     |             |                               |
id-ref-installed           |            | Installed   |                     |             |                               |
id-ref-extend-main         |            | Imported    |                     |             |                               |
augm_leafref_m1            |            | Installed   | root:root           | 666         |                               |
augm_leafref_m2            |            | Implemented | root:root           | 666         |                               |
module-a                   | 2016-02-10 | Installed   | root:root           | 666         |  sub-a-one sub-a-two          |
module-b                   | 2016-02-05 | Installed   | root:root           | 666         |                               |
module-a                   | 2016-02-02 | Imported    |                     |             |  sub-a-one sub-a-two          |
state-module               | 2016-07-01 | Installed   | root:root           | 666         |                               |
top-level-mandatory        |            | Installed   | root:root           | 666         |                               |
cross-module               |            | Installed   | root:root           | 666         |                               |
turing-machine             | 2013-12-27 | Installed   | root:root           | 666         |                               |
ietf-netconf-notifications | 2012-02-06 | Installed   | root:root           | 666         |                               |
ietf-netconf               | 2011-06-01 | Installed   | root:root           | 666         |                               | writable-running candidate rollback-on-error validate startup xpath
nc-notifications           | 2008-07-14 | Installed   | root:root           | 666         |                               |
notifications              | 2008-07-14 | Installed   | root:root           | 666         |                               |
servers                    |            | Installed   | root:root           | 666         |                               |
commit-nacm                |            | Installed   | root:root           | 666         |                               |
swig-test-cpp-changes      | 2017-03-09 | Installed   | root:root           | 666         |                               |
swig-test-cpp-operations   | 2017-03-09 | Installed   | root:root           | 666         |                               |
ietf-netconf-with-defaults | 2011-06-01 | Installed   |                     |             |                               |
ietf-netconf-monitoring    | 2010-10-04 | Installed   | root:root           | 666         |                               |
ietf-yang-library          | 2018-01-17 | Installed   | root:root           | 666         |                               |
ietf-datastores            | 2017-08-17 | Imported    |                     |             |                               |
```
Then I installed all your modules (from the attached files):
```
sysrepoctl -i -g issue/routing-indra@2018-09-18.yang
sysrepoctl -i -g issue/ietf-ip@2014-06-16.yang
sysrepoctl -i -g issue/iana-if-type.yang
```
which added this output to sysrepo modules
```
routing-indra              | 2018-09-18 | Installed   | root:root           | 666         |                               |
ietf-ip                    | 2014-06-16 | Installed   |                     |             |                               |
ietf-interfaces            | 2014-05-08 | Implemented | root:root           | 666         |                               |
ietf-routing               | 2015-05-25 | Implemented | root:root           | 666         |                               |
iana-if-type               | 2014-05-08 | Installed   |                     |             |                               |
```
Now I started, in order, `sysrepod`, `application_example ietf-interfaces`, `netopeer2-server`, and `netopeer2-cli`. Using the CLI I connected and then executed this RPC successfully
```
edit-config --target running --config=issue/ipv6_merge.xml
```
My application example full output was
```
application_example ietf-interfaces


 ========== READING STARTUP CONFIG ietf-interfaces: ==========

/ietf-interfaces:interfaces (container)


 ========== STARTUP CONFIG ietf-interfaces APPLIED AS RUNNING ==========



 ========== CONFIG HAS CHANGED, CURRENT RUNNING CONFIG ietf-interfaces: ==========

/ietf-interfaces:interfaces (container)
/ietf-interfaces:interfaces/interface[name='eth0'] (list instance)
/ietf-interfaces:interfaces/interface[name='eth0']/name = eth0
/ietf-interfaces:interfaces/interface[name='eth0']/type = iana-if-type:ethernetCsmacd
/ietf-interfaces:interfaces/interface[name='eth0']/enabled = true
/ietf-interfaces:interfaces/interface[name='eth0']/ietf-ip:ipv6 (container)
/ietf-interfaces:interfaces/interface[name='eth0']/ietf-ip:ipv6/address[ip='fd00::62'] (list instance)
/ietf-interfaces:interfaces/interface[name='eth0']/ietf-ip:ipv6/address[ip='fd00::62']/ip = fd00::62
/ietf-interfaces:interfaces/interface[name='eth0']/ietf-ip:ipv6/address[ip='fd00::62']/prefix-length = 64
/ietf-interfaces:interfaces/interface[name='eth0']/ietf-ip:ipv6/forwarding = false
/ietf-interfaces:interfaces/interface[name='eth0']/ietf-ip:ipv6/enabled = true
/ietf-interfaces:interfaces/interface[name='eth0']/ietf-ip:ipv6/dup-addr-detect-transmits = 1
/ietf-interfaces:interfaces/interface[name='eth0']/ietf-ip:ipv6/autoconf (container)
/ietf-interfaces:interfaces/interface[name='eth0']/ietf-ip:ipv6/autoconf/create-global-addresses = true [default]
```
Can you try it or tell me what are you doing differently?

Regards,
Michal

**Comment by vhuertas** [comment]
Dear Michal,
I have tried exactly what you have done step-ty-step and no error was displayed in the merge. However there is one missing step you haven't done that is what allows you to reproduce the issue:

Before starting you application and the netopeer2-server you must run:

**sudo sysrepoctl --feature-enable=routing --module=routing-indra**

This enables the container where "max-ip-routes" and other leafs with mandatory="true" are.
Then you start your application and the netopeer2-server.
When you issue the Netopeer-cli command:

**edit-config --target running --config=issue/ipv6_merge.xml**

and the following error is displayed:

        ERROR
        type:     application
        tag:      operation-failed
        severity: error
        path:     /routing-indra:routing-service-config/license-params
        message:  Missing required element "max-ip-routes" in "license-params".

        type:     application
        tag:      operation-failed
        severity: error
        path:     /routing-indra:routing-service-config/license-params
        message:  Missing required element "max-ip-routes" in "license-params".

Try to do it and see if this error is displayed as well in your environment.

Thanks again for your attention.

NOTE: in the test I performed, I installed using "sysrepoctl --install" command the ietf-interfaces, ietf-ip, ietf-yang-types, ietf-inet-types, iana-if-type, ietf-ipv4-unicast-routing, ietf-ipv6-unicast-routing, ietf-event-notifications and ietf-routing. You can see subsequently the current status of 'sysrepoctl -l':

Module Name                | Revision   | Conformance | Data Owner          | Permissions | Submodules                    | Enabled Features
-----------------------------------------------------------------------------------------------------------------------------------------------
ietf-netconf-notifications | 2012-02-06 | Installed   | root:root           | 666         |                               |
ietf-netconf               | 2011-06-01 | Installed   | root:root           | 666         |                               | writable-running candidate rollback-on-error validate startup xpath url
ietf-netconf-acm           | 2012-02-22 | Imported    |                     |             |                               |
nc-notifications           | 2008-07-14 | Installed   | root:root           | 666         |                               |
notifications              | 2008-07-14 | Installed   | root:root           | 666         |                               |
ietf-netconf-with-defaults | 2011-06-01 | Installed   |                     |             |                               |
ietf-netconf-monitoring    | 2010-10-04 | Installed   | root:root           | 666         |                               |
ietf-yang-library          | 2018-01-17 | Installed   | root:root           | 666         |                               |
ietf-yang-types            | 2013-07-15 | Installed   |                     |             |                               |
turing-machine             | 2013-12-27 | Installed   | root:root           | 666         |                               |
ietf-datastores            | 2017-08-17 | Installed   |                     |             |                               |
ietf-inet-types            | 2013-07-15 | Installed   |                     |             |                               |
ydk                        | 2016-02-26 | Installed   | root:root           | 666         |                               |
ietf-qos-policy            | 2016-03-03 | Installed   | root:root           | 666         |                               |
ietf-qos-classifier        | 2016-03-03 | Installed   | root:root           | 666         |                               |
ietf-qos-action            | 2016-06-15 | Installed   | root:root           | 666         |                               |
ietf-access-control-list   | 2016-10-12 | Installed   | root:root           | 666         |                               |
ietf-packet-fields         | 2016-10-12 | Installed   |                     |             |                               |
routing-indra              | 2018-09-18 | Installed   | root:root           | 666         |                               | routing
ietf-ip                    | 2014-06-16 | Installed   |                     |             |                               |
ietf-interfaces            | 2014-05-08 | Installed   | root:root           | 666         |                               | if-mib
ietf-routing               | 2015-10-16 | Installed   | root:root           | 666         |                               |
iana-if-type               | 2014-05-08 | Installed   |                     |             |                               |
ietf-ipv4-unicast-routing  | 2015-10-16 | Installed   |                     |             |                               |
ietf-ipv6-unicast-routing  | 2015-10-16 | Installed   |                     |             |                               |
tcp-acceleration-indra     | 2018-09-18 | Installed   | root:root           | 666         |                               |
smart-routing-indra        | 2018-09-18 | Installed   | root:root           | 666         |                               |
qos-indra                  | 2018-09-18 | Installed   | root:root           | 666         |                               |
ietf-event-notifications   | 2016-10-27 | Installed   | root:root           | 666         |                               |

**Comment by michalvasko** [comment]
Hi,
you are right, that was the problem. Anyway, it all works as expected albeit not perfectly. The issue is that for *sysrepo* `routing-indra` data depend on `ietf-interfaces` data. That is actually not true but fixing that would be a significant change that will simply not happen.

Keeping this in mind, when you change `ietf-interfaces` data with `ipv6_merge.xml`, also `routing-indra` data are validated and they correctly fail. There is another implicit issue in your case, when you enable the `routing` feature, your stored (empty) `routing-indra` data are no longer valid. *sysrepo* should handle this better (not even allow it) but it only warns about it in the output. So, if you make the data valid, everything should work correctly.

Regards,
Michal

**Comment by vhuertas** [comment]
Dear Michal,

In my application I use sysrepo C API to initialize the routing-indra data (running datastore) as soon as it starts. So, taking this into account, the data is already valid and the sysrepo should find it in the running datastore when I send the merge for the IPv6 address configuration. So this data validation should be OK, isn't it?

Will this bugfix be included in the master branch so that I can recompile it and reinstall it?

Anyway, in order to avoid any further issue with other models defined by me (and have augments to ietf-interfaces and other ietf models) I will have to remove the mandatory=true from each leaf. This is an impact I can deal with in my application. This doesn't mean that this bugfix is irrelevant as someone else may find the same issue someday.

Regards

**Comment by michalvasko** [comment]
Hi,
> In my application I use sysrepo C API to initialize the routing-indra data (running datastore) as soon as it starts.

What exactly do you mean, right after you subscribe to its data, you configure the mandatory leaves in `<running>`? Then you should not be having any problems.

> Will this bugfix be included in the master branch so that I can recompile it and reinstall it?

There is no bugfix, I have removed the last referenced commit because it actually makes no real difference.

Regards,
Michal

**Comment by vhuertas** [comment]
Hi,

No. I configure the mandatory data and after this I subscribe to running config. Doing so it worked as when I issue a <get-config> the netopeer returns what I have configured.

Regards,

**Comment by michalvasko** [comment]
You configure the mandatory data in `<startup>`? I am going to assume so because otherwise you would get error that you are configuring non-enabled data. In that case, once you subscribe they should get copied to `<running>` and hence be valid. I am quite confused now, could you please elaborate a bit more because this does not make sense.

Regards,
Michal

**Comment by vhuertas** [comment]
Hi Michal,

I do it this way at startup right from the begining:

1. I connect to sysrepo (sr_connect): (connection only used for ietf-interfaces model)
2. I establish a session with sysrepo to STARTUP (sr_session_start to datastore SR_DS_STARTUP): (Session only used for ietf-interfaces model)

3. Establish a second connection (sr_connect): (connection only used for indra-routing model)
4. I establish a session with sysrepo to STARTUP (sr_session_start to datastore SR_DS_STARTUP): (Session only used for ' indra-routing' model)

5. Using session in step 2 (STARTUP) I clean all the info contained because I want to make sure that at startup no remaining config is applied (sr_delete-item)
6. Perform a commit of the STARTUP session to apply the clean order (sr_commit)
7. Configure all the ietf-interfaces params I want to set (mandatory and non-mandatory) using the session in step 2 (STARTUP): sr_set_item_str.
8. Perform a commit of the STARTUP session to apply the parameters configured in the previous step (sr-commit)

9. Repeat exactly the steps 5 to 8 but using the session in step 4 in order to configure mandatory data in 'routing-indra' model.

10. unsubscribe from STARTUP session (sr_unsubscribe) and close session established in step 2 ( for ietf-interfaces): sr_session_stop
11. start a new session (sr_connect and sr_session_start) with RUNNING data store for ietf-interfaces model.
12. subscribe to the /ietf-interfaces:interfaces-state (RUNNING session): (sr_dp_get_items_subscribe with SR_SUBSCR_DEFAULT)
13. subscribe to the /ietf-interfaces in case that mofidications are preformed from NETCONF (sr_module_change_subscribe with SR_SUBSCR_DEFAULT)

14. repeat steps 11 to 13 (new connection and session with RUNNING only for 'routing-indra' model).

Regards

**Comment by michalvasko** [comment]
Okay (thanks for the details), so after this you try to apply `ipv6_merge.xml` and it fails with the validation error? Second question, what is the `<running>` configuration after all these steps for `indra-routing`?

Regards,
Michal

**Comment by vhuertas** [comment]
> Okay (thanks for the details), so after this you try to apply `ipv6_merge.xml` and it fails with the validation error? 

Yes. Of course the error occurs once "routing" feature is enabled in routing-indra model.

>Second question, what is the `<running>` configuration after all these steps for `indra-routing`?

If the 'routing' feature is enabled the result is this using Netopeer2-cli:
 get-config --source running --filter-xpath=/routing-indra:routing-service-config
DATA
```
<routing-service-config xmlns="http://qosenforcer.indra.es/routing">
  <license-params>
    <enabled-service>false</enabled-service>
    <max-ip-routes>1000</max-ip-routes>
    <max-wan-profiles>10</max-wan-profiles>
    <max-ip-flows>100000</max-ip-flows>
  </license-params>
</routing-service-config>
```
If the feature is not enabled no configuration is retrieved.

Regards,

**Comment by michalvasko** [comment]
Hi,
so please try the latest *sysrepo* `devel`, your use-case should work and there should be fewer error and warning messages.

Regards,
Michal

**Comment by vhuertas** [comment]
Dear Michal,

I have updated sysrepo (devel), libyang and libnetconf2.
sysrepo compiles fine but when executing "make install" an error occurs and I am stuck here:

Install the project...
-- Install configuration: "Release"
-- Up-to-date: /etc/sysrepo
-- Up-to-date: /etc/sysrepo/yang/
-- Up-to-date: /etc/sysrepo/data/
-- Up-to-date: /etc/sysrepo/data/internal/
-- Up-to-date: /var/run/sysrepo-subscriptions
-- Up-to-date: /usr/lib64/sysrepo/plugins
-- Up-to-date: /etc/sysrepo/yang/internal/sysrepo-persistent-data.yang
-- Up-to-date: /etc/sysrepo/yang/internal/sysrepo-module-dependencies.yang
-- Up-to-date: /etc/sysrepo/yang/internal/sysrepo-notification-store.yang
-- Up-to-date: /usr/lib64/pkgconfig/libsysrepo.pc
-- Up-to-date: /usr/share/man/man1/sysrepoctl.1
-- Up-to-date: /usr/share/man/man1/sysrepocfg.1
-- Exec: /home/cuda1/git/sysrepo/build/src/sysrepoctl --install --yang=/home/cuda1/git/sysrepo/yang/ietf-netconf-notifications.yang --permissions=666
-- Exec: /home/cuda1/git/sysrepo/build/src/sysrepoctl --install --yang=/home/cuda1/git/sysrepo/yang/nc-notifications.yang --permissions=666
-- Exec: /home/cuda1/git/sysrepo/build/src/sysrepoctl --install --yang=/home/cuda1/git/sysrepo/yang/notifications.yang --permissions=666
-- Installing: /usr/lib64/libsysrepo.so.0.7.7
-- Up-to-date: /usr/lib64/libsysrepo.so.0.7
-- Up-to-date: /usr/lib64/libsysrepo.so
-- Set runtime path of "/usr/lib64/libsysrepo.so.0.7.7" to ""
-- Installing: /usr/lib64/libsysrepo.a
-- Installing: /usr/bin/sysrepod
-- Set runtime path of "/usr/bin/sysrepod" to ""
-- Installing: /usr/bin/sysrepo-plugind
-- Set runtime path of "/usr/bin/sysrepo-plugind" to ""
-- Installing: /usr/bin/sysrepoctl
-- Set runtime path of "/usr/bin/sysrepoctl" to ""
-- Installing: /usr/bin/sysrepocfg
-- Set runtime path of "/usr/bin/sysrepocfg" to ""
-- Up-to-date: /usr/include
-- Up-to-date: /usr/include/sysrepo
-- Installing: /usr/include/sysrepo/plugins.h
-- Installing: /usr/include/sysrepo/xpath.h
-- Installing: /usr/include/sysrepo/values.h
-- Installing: /usr/include/sysrepo/trees.h
-- Installing: /usr/include/sysrepo.h
-- Exec: /home/cuda1/git/sysrepo/build/src/sysrepoctl --install --yang=/home/cuda1/git/sysrepo/examples/yang/turing-machine.yang --permissions=666; /home/cuda1/git/sysrepo/build/src/sysrepocfg --import=/home/cuda1/git/sysrepo/examples/yang/turing-machine.data.xml --datastore=startup --format=xml --level=0 turing-machine
-- Exec: /home/cuda1/git/sysrepo/build/src/sysrepoctl --install --yang=/home/cuda1/git/sysrepo/examples/yang/iana-if-type.yang --permissions=666
-- Exec: /home/cuda1/git/sysrepo/build/src/sysrepoctl --install --yang=/home/cuda1/git/sysrepo/examples/yang/ietf-ip@2014-06-16.yang --permissions=666
-- Exec: /home/cuda1/git/sysrepo/build/src/sysrepoctl --install --yang=/home/cuda1/git/sysrepo/examples/yang/ietf-interfaces@2014-05-08.yang --permissions=666; /home/cuda1/git/sysrepo/build/src/sysrepocfg --import=/home/cuda1/git/sysrepo/examples/yang/ietf-interfaces.data.xml --datastore=startup --format=xml --level=0 ietf-interfaces
CMake Error at examples/cmake_install.cmake:68 (message):
  Error: 1
Call Stack (most recent call first):
  cmake_install.cmake:167 (include)


Makefile:85: recipe for target 'install' failed
make: *** [install] Error 1

Any idea?

Regards,

**Comment by michalvasko** [comment]
Hi,
this happened to me as well sometimes but I never investigated further because it always got fixed by clearing the whole build directory and running the compilation from scratch. I suggest you do the same.

Regards,
Michal
